### PR TITLE
feat: redstone oracle adapter test

### PR DESCRIPTION
**Pull request by md0x**
Proposed Changes in this PR:

- Test the integration of a RedStone Classic oracle with Oval and ChainlinkSourceAdapter.
- Utilize the MergedPriceFeedAdapterWithRounds smart contract for the testing.
- Fetch prices from the RedStone production gateway at https://oracle-gateway-1.a.redstone.finance/ and submit them to the adapter.
- Demonstrate that Oval requires no modifications to work seamlessly with RedStone Classic.

### CODE REVIEW

File: test/unit/RedStoneOracle.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            redstonePayload
+        );
+
+        address(redstoneOracle).call(encodedFunctionWithRedstonePayload);
+
+        return updatePrice;
+    }
+
+    function testPushPrice() public {
+        vm.warp(1234);
```
Comment (md0x) [inline test/unit/RedStoneOracle.sol:66]: The timestamp associated to the price pushed is the `block.timestamp` when the price is pushed.

File: test/unit/RedStoneOracle.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+contract RedstoneOracleAdapterTest is CommonTest, MockRedstonePayload {
+    RedstonePriceFeedWithRounds redstoneOracle;
+    TestedSourceAdapter sourceAdapter;
+
+    function setUp() public {
+        redstoneOracle = new RedstonePriceFeedWithRounds(bytes32("BTC"));
+        sourceAdapter = new TestedSourceAdapter(
+            IAggregatorV3Source(address(redstoneOracle))
```
Comment (chrismaree) [inline test/unit/RedStoneOracle.sol:38]: this is great! we can show that we can import redstone sources, treat them exactly as chainlink, and everything continues to work as we'd expect.

File: test/unit/RedStoneOracle.sol (hunk 1 above)
Comment (chrismaree) [reply test/unit/RedStoneOracle.sol:66]: yup, same as with chainlink.

**Review by chrismaree — COMMENTED** [review]
This is awesome. it shows we can use Redstone with zero changes, treating it directly as chainlink. only downside with this PR right now is that it has a lot of copy paste files to get the deployments of tests working. I'll try create a follow on commit that moves this into more minimal imports.

**Review by Reinis-FRP — APPROVED** [review]
nice, and agree with @chrismaree to clean dependencies in follow-up

**Review by mrice32 — APPROVED** [review]
LGTM!

File: test/fork/adapters/ChainlinkSourceAdapter.sol [hunk 3]
```diff
@@ -45,26 +45,26 @@ contract ChainlinkSourceAdapterTest is CommonTest {
         uint256 targetTime = block.timestamp - 1 hours;
         (uint80 latestRound,,,,) = chainlink.latestRoundData();
         (int256 lookBackPrice, uint256 lookBackTimestamp) = sourceAdapter.tryLatestDataAt(targetTime, 10);
-        (, int256 answer, uint256 startedAt,,) = chainlink.getRoundData(latestRound - 1);
-        assertTrue(startedAt <= targetTime); // The time from the chainlink source is at least 1 hours old.
+        (, int256 answer,, uint256 updatedAt,) = chainlink.getRoundData(latestRound - 1);
```
Comment (chrismaree) [inline test/fork/adapters/ChainlinkSourceAdapter.sol:48]: after writing the other tests (for redstone) I realized this is inconsistant with qwhat the adapter actually dones, so updated it.

**Review by chrismaree — COMMENTED** [review]
I updated this PR somewhat (mainnet fork tests, moved scripts. toethers, moved scripts to typescript, cleaned up. someof the imports and pathing). but I'm still not totally done. I don't like that we have toe `lib`  as a directory of files, it should be a git submodule. This is currently broken on the redstone imports. andneeds a bit of effort to fix. we should address this before merging this PR.

File: scripts/getRedstonePayload.log.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+You have to provide a data FeedAn error occurred: Request failed {"reqParams":{"dataServiceId":"redstone-primary-prod","uniqueSignersCount":3,"dataFeeds":["--no-warnings"],"urls":["https://oracle-gateway-1.a.redstone.finance"]}}, Original error: AggregateError: , errors: Error: Requested data feed id is not included in response: --no-warnings
```
Comment (chrismaree) [inline scripts/getRedstonePayload.log.txt:1]: This should not be committed and given its in the .gitignore it wont be in the future.

File: remappings.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
 ds-test/=lib/forge-std/lib/ds-test/src/
 forge-std/=lib/forge-std/src/
-openzeppelin-contracts/=lib/openzeppelin-contracts/
+redstone-oracles-monorepo/=lib/redstone-contracts/src/RedStoneBaseContracts/redstone-oracles-monorepo
```
Comment (chrismaree) [inline remappings.txt:3]: do you think naming it like `redstone-oracles-monorepo` makes sense? prehaps `redstone-oracle`?

File: remappings.txt [hunk 6]
```diff
@@ -1,3 +1,3 @@
 ds-test/=lib/forge-std/lib/ds-test/src/
 forge-std/=lib/forge-std/src/
-openzeppelin-contracts/=lib/openzeppelin-contracts/
```
Comment (chrismaree) [inline remappings.txt:3]: also what happened here?

**Review by chrismaree — COMMENTED** [review]
I think this is great. the only thing that once could argue we don't want is the file `src/oracles/RedstonePriceFeedWithRounds.sol`. this _could_ be pulled into the test where its used? otherwise this file has no other direct usage and is a bit different to the rest of the repo?

File: src/adapters/source-adapters/ChainlinkSourceAdapter.sol [hunk 7]
```diff
@@ -6,6 +6,7 @@ import {SafeCast} from "openzeppelin-contracts/contracts/utils/math/SafeCast.sol
 import {DecimalLib} from "../lib/DecimalLib.sol";
 import {IAggregatorV3Source} from "../../interfaces/chainlink/IAggregatorV3Source.sol";
 import {DiamondRootOval} from "../../DiamondRootOval.sol";
+import "forge-std/console.sol";
```
Comment (chrismaree) [inline src/adapters/source-adapters/ChainlinkSourceAdapter.sol:9]:
```suggestion
```

File: src/oracles/RedstonePriceFeedWithRounds.sol [hunk 8]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.17;
+
```
Comment (chrismaree) [inline src/oracles/RedstonePriceFeedWithRounds.sol:3]: if this is going to live in the repo here it needs comments & natspec

Comment (md0x) [reply src/oracles/RedstonePriceFeedWithRounds.sol]: @chrismaree I pulled this out to https://github.com/UMAprotocol/redstone-contracts/tree/master/src/examples

File: remappings.txt (hunk 6 above)
Comment (md0x) [reply remappings.txt:3]: I realized it wasn't needed when I was cleaning up the remappings.
Everything is imported with `"openzeppelin-contracts/..."` and Foundry automatically reads from the lib folder.

**Review by chrismaree — APPROVED** [review]
This is great! love showing support for something by simple adding some tests :)

**Pull request merged**